### ISSUE TITLE
displayDialogAsync and messageParent functions aren't working correctly when messaging across domains

### ISSUE THREAD

**Issue by Feetterer20** [OfficeDev/office-js#2395]
# Provide required information needed to triage your issue

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: PC desktop
* Host [Excel, Word, PowerPoint, etc.]: Word & Excel
* Office version number: Version 2108 (Build 14326.20748 Click-to-Run)
* Operating System: Windows 10 Enterprise, build 19042.1348
* Browser (if using Office on the web): N/A
 
## Expected behavior

When utilizing `Office.context.ui.messageParent("closed", { targetOrigin: "*" };` in an open Dialog I was expecting the "closed" message to be picked up by the `Office.EventType.DialogMessageReceived` event handler in the Office Add-In. The Dialog was opened using `Office.context.ui.displayDialogAsync`, but this event handler isn't firing when a message is sent.

## Current behavior

Using the functions and event handlers mentioned above, the `Office.EventType.DialogMessageReceived` event handler is never fired when we call `Office.context.ui.messageParent("closed", { targetOrigin: "*" };`

I think it's important to note that we're trying to send messages across domains (in this case when troubleshooting this in a local development environment, this is two different local host domains). When we were troubleshooting this, I tested out same-domain messaging using the same functions mentioned above, and that worked as expected. The issue seems to arise when we want to send messages across domains.

It's also worth noting that the same code was working fine (messaging across domains) when it was initially implemented back in August of 2021. Since then, it seems that things stopped working without us making any updates. 

## Steps to reproduce

We have an Office Task Pane Add-In that we created to be used in Word and Excel files. This add in has buttons which are responsible for opening a Dialog that contains a custom form (MVC Web App Controller Action & View) which the user fills out and then clicks a custom "Save and Close" button which is responsible for calling the `Office.context.ui.messageParent("closed", { targetOrigin: "*" };` function to ultimately close the Dialog. Again, I think it's important to note that the MVC page and Office Add In are on different domains, so we're looking to utilize cross-domain messaging.

JavaScript code from the Office Add-In (which is called the Work Panel in our solution). This function is called when the user clicks one of the custom buttons on the add-in, and this opens the custom form (MVC Web App Controller Action & View):
```
var dialog;
function OpenPopup({ hyperLink, doRefresh, isBrowserWindow,
    customHeightPercentage, customWidthPercentage }) {

    var customHeight = customHeightPercentage ? customHeightPercentage : (screen.availWidth * 0.95);
    var customWidth = customWidthPercentage ? customWidthPercentage : (screen.availHeight * 0.85);
    if (Boolean.parse(isBrowserWindow)) {
        Office.context.ui.openBrowserWindow(hyperLink);
    }
    else {
        Office.context.ui.displayDialogAsync(hyperLink, { height: parseInt(customHeight), width: parseInt(customWidth) }, (asyncResult) => {

             if (asyncResult.status === Office.AsyncResultStatus.Failed) {
                var message = JSON.stringify(asyncResult);
                ShowToastMessage("error", "Error displayDialogAsync: " + message, "Error", 5000);
                return;
            }

            dialog = asyncResult.value;
            dialog.addEventHandler(Office.EventType.DialogEventReceived, function (args) {
                if (doRefresh && (args.message === 'closed' || args.error === 12006)) {
                    refreshTab();
                    dialog.close();
                }
            });
            dialog.addEventHandler(Office.EventType.DialogMessageReceived, function (args) {
                if (doRefresh && args.message === 'closed') {
                    refreshTab();
                    dialog.close();
                }
            });
        });
    }
}
```

JavaScript code from the MVC Web App (separate domain from the Add-In), which is called when the user clicks a "Save and Close" button on the custom form:
```
function returnToRedirectUrl() {
    // If the popup is opened from the workpanel then we want to sent message to the host page.
    var fromOfficeDialog = decodeURIComponent(getQueryStringParameter('fromOfficeDialog'));
    if (fromOfficeDialog.startsWith("true")) {
        Office.context.ui.messageParent("closed", { targetOrigin: "*" });
    }
}
```

# Provide additional details

During the troubleshooting of this issue, I've seen that the Dialog Origin 1.1 requirement set is necessary for cross-domain messaging to work correctly. 

Something worth noting is that when calling `Office.context.requirements.isSetSupported("DialogOrigin", "1.1");` with the Add-In JavaScript code, we get a `true` result, but when calling the same function with the MVC Web App JavaScript code, we get a `false` result. 

## Context

We are using the `Microsoft.Office.js v1.1.0.18` Nuget package in both the Office Add In and other MVC Web App projects.

With the messaging not working as expected, this causes our "Save and Close" button to not properly close the Dialog for our users. If needed, I would be open to any alternatives that we can use with Office JS to accomplish this same goal, but we are hoping that we can utilize the functions mentioned above to maintain this functionality for our users.

Please let me know if any further information / context is needed to help in any triaging. Thank you for your help!

**Comment by pkkj** [comment]
Hi,

There is one more step that can help us to triage the problem in your environment. Would you like to try to side load this [add-in](https://kepengtestapp.azurewebsites.net/manifest/AzureDialogTest.xml) in your test environment? You could examine the source code of the [main page](https://kepengtestapp.azurewebsites.net/pages/dialog-taskpane.html) before you load it. There are several button in the page and it can launch a dialog in different domain. There is button in the dialog so you could test the MessageParent API.

Thanks,
Kejing

**Comment by Feetterer20** [comment]
@pkkj Thank you for the quick response! I ran your add-in in my local testing environment and utilized the cross-domain messaging and it appears to be working as expected. 

Some things I noticed in your code that is different from our solution:
 - When I called `Office.context.requirements.isSetSupported("DialogOrigin", "1.1")` through the developer tools for both the add-in and the separate dialog code, it returned `true` for both of them for your code. For our solution that we're having troubles with, I was receiving `true` for the add-in but `false` on the dialog side.
 - I noticed you're loading the Office.js code from the CDN, where we're using the `Microsoft.Office.js v1.1.0.18` Nuget package.
 - It looks like you're adding the `Office.EventType.DialogParentMessageReceived` event handler in both the Add-in and Dialog code, is this event handler necessary in the Dialog code?

Thank you again for your help!

**Comment by pkkj** [comment]
Hi. I am wondering is it possible to use the CDN office.js in your environment for your add-in and verify whether the API is working. The Office.js is being updated very frequently so the older version might not work correctly. And please make sure that both your taskpane and dialog are consuming the Office.js from CDN.

Regarding the `Office.EventType.DialogParentMessageReceived`, I need this in dialog but I don't think I use it in the taskpane. Would you like to double check?

Thanks,
Kejing

**Comment by Feetterer20** [comment]
@pkkj After updating my local environment to consume `Office.js` from the CDN, I was able to verify that messages were being sent correctly across domains and the code in our solution was then working as expected.

With regards to where we need to add the `Office.EventType.DialogParentMessageReceived` event handler, I was able to get things working correctly with just having this added to the taskpane.

From this it seems like the `Office.js` code included in the `Microsoft.Office.js v1.1.0.18` Nuget package may just be too out of date. Our hesitation with consuming this code from the CDN in production is that we don't want our Add-In to stop working unexpectedly due to any updates in the CDN. Are there any newer versions of this Nuget package that we would be able to download and use in our solution (or any alternatives that you would recommend)?

Thank you very much for your help, I really do appreciate it.

**Comment by Feetterer20** [comment]
@pkkj To follow up on this issue, we saw there was a new `Microsoft.Office.js` Nuget package released (`v1.1.0.19`) which we have installed in our solution and this has resolved our issue.

Thank you again for all of your help, it is very much appreciated!
